addons/custom-block-shape/update-all-blocks.js

```diff
@@ -1,24 +1,20 @@
-export function updateAllBlocks(vm, workspace, blockly) {
-  const eventsOriginallyEnabled = blockly.Events.isEnabled();
-  blockly.Events.disable(); // Clears workspace right-click→undo (see SA/SA#6691)
+export function updateAllBlocks(blockly, workspace, { updateFlyout = true } = {}) {
+  // Calling Events.disable() will:
+  // - prevent changes to the workspace from being added to the undo stack;
+  // - prevent project change events from being triggered.
+  // From scratch-blocks source code: "Every call to this function MUST also call enable."
+  blockly.Events.disable();
 
   if (workspace) {
-    if (vm.editingTarget) {
-      vm.emitWorkspaceUpdate();
-    }
+    blockly.Xml.clearWorkspaceAndLoadFromXml(blockly.Xml.workspaceToDom(workspace), workspace);
     const flyout = workspace.getFlyout();
-    if (flyout) {
+    if (updateFlyout && flyout) {
       const flyoutWorkspace = flyout.getWorkspace();
-      window.Blockly.Xml.clearWorkspaceAndLoadFromXml(
-        window.Blockly.Xml.workspaceToDom(flyoutWorkspace),
-        flyoutWorkspace
-      );
+      blockly.Xml.clearWorkspaceAndLoadFromXml(blockly.Xml.workspaceToDom(flyoutWorkspace), flyoutWorkspace);
       workspace.getToolbox().refreshSelection();
-      workspace.toolboxRefreshEnabled_ = true;
     }
+    workspace.toolboxRefreshEnabled_ = true;
   }
 
-  // There's no particular reason for checking whether events were originally enabled.
-  // Unconditionally enabling events at this point could, in theory, cause bugs in the future.
-  if (eventsOriginallyEnabled) blockly.Events.enable(); // Re-enable events
+  blockly.Events.enable();
```

Comment:
Calling `enable()` unconditionally is correct: if `disable()` was called while events were disabled, `enable()` needs to be called twice to actually reenable them. See https://github.com/scratchfoundation/scratch-blocks/blob/1421093/core/events.js#L282-L303

Reply:
Thanks. I think I wrote that code comment. As I am not fully familiar with the scratch-blocks codebase, I didn't want to risk it.